### ISSUE TITLE
Exact formulas

### ISSUE THREAD

**Issue by MaxHalford** [jrzaurin/LightGBM-with-Focal-Loss#3]
Hey there! Just wanted to let you know I put some work into deriving exact formulas, which avoids the need for numeric differentiation. I also found out that to be correct, the initialization value of the gradient boosting algorithm has to be accounted for. I've written about this in [a blog post](https://maxhalford.github.io/blog/lightgbm-focal-loss/).

Kind regards :)

**Comment by jrzaurin** [comment]
Hey @MaxHalford 

Thanks! I will have a look to your post for sure! 

I must admit I gave it a go but got a bit tedious, moreover when considering multi-class problems. Looking forward to read the post really. 

Regarding the initialization value, it is true, one should account not only for the focal loss, but for any loss. What happened for what I remember is that 1) digging into it was not trivial at all and 2) the empirical results were really good with 0 (the default). Therefore I did not bother (maybe I should). 

Maybe you could consider contributing to the repo and we could add it (?)

Anyway, I will read the post asap! I would be interested to see the improvement in running time. Thanks for letting me know and for the work doing derivatives 🙂

Cheers

**Comment by MaxHalford** [comment]
Thanks for your kind comment.

> Regarding the initialization value, it is true, one should account not only for the focal loss, but for any loss. What happened for what I remember is that 1) digging into it was not trivial at all and 2) the empirical results were really good with 0 (the default). Therefore I did not bother (maybe I should).

As you'll see in the post, I gave up on the maths and ended up using numeric optimisation. This is a cheap win because most loss functions are convex!

> Maybe you could consider contributing to the repo and we could add it (?)

I would be inclined to contribute to the repo. I think there's also some potential to make a more generic library devoted to a wider range of loss functions that are not implemented in LightGBM, XGBoost, CatBoost, etc.
